cosmos/job/drm/drm_ge.py

```diff
@@ -301,17 +301,88 @@ def qacct(job_id, num_retries=10, quantum=30, logger=None, log_prefix=""):
     return good_qacct_dict if good_qacct_dict else curr_qacct_dict
 
 
-def qstat():
+def qdel(job_ids, logger):
+    """
+    Call qdel on all the supplied job_ids: if that fails, qdel each job_id individually.
+
+    Unlike other SGE cli commands, each qdel call is attempted only once, with a
+    fairly harsh one-second timeout, because function method is often called in
+    an exit handler that does not have arbitrary amounts of time in which to run.
+    """
+    stdout, stderr, returncode = run_cli_cmd(
+        ["qdel", ",".join(job_ids)],
+        logger=logger,
+        attempts=1,
+        timeout=1,
+    )
+    if returncode == 0:
+        logger.info("qdel reported successful signalling of %d job_ids", len(job_ids))
+        return len(job_ids)
+
+    successful_qdels = 0
+    undead_job_ids = []
+
+    for job_id in job_ids:
+        if "has deleted job %s" % job_id in stdout:
+            successful_qdels += 1
+        elif "has registered the job %s for deletion" % job_id in stdout:
+            successful_qdels += 1
+        else:
+            undead_job_ids.append(job_id)
+
+    if undead_job_ids:
+        logger.warning(
+            "qdel returned exit code %s, calling on one job_id at a time", returncode
+        )
+
+        for job_id in undead_job_ids:
```

Comment:
what are the pros/cons of using a ThreadPoolExecutor here? pros -- would return more quickly. cons -- would hammer grid engine at a time when we're not sure it's working 100%

Reply:
maybe add a comment to that effect